jkglasbrenner/ising-model-py
# Working Python implementation of 2D Ising model

* Python code heavily inspired by my Javascript implementation
  https://github.com/jkglasbrenner/ising-model-js
* Development toolchain
  * Add attrs to toolchain
  * Updated toolchain to hook into pytest and install pytest-runner
  * Execute pytest unit tests from Makefile
  * Update help message in Makefile
  * Install Sphinx plugin to use typehints
  * Update pip
* Simulation modules
  * algorithms.py: Implementation of Metropolis algorithm
  * data.py: Data containers for storing state and simulation history
  * distributions.py: State and probability distributions
  * model.py: Specifics of Ising model
  * run.py: Interface for running the simulations
  * sampling.py: How sites are sampled and sweeped during simulation
  * statistics.py: How estimators statistics are computed
  * Allow easy import of modules within isingmodel package
* Pytest-based unit tests
  * Add unit tests for BinaryLattice class to test that site generation and
    neighbor finding is correct.
  * Unit test that tests full simulation as a short run
  * Test ferromagnetic simulation with Neumann and Moore neighborhoods
  * Check that final energy and magnetization fall within the expected upper and
    lower bounds
  * Generate random seed so that the test runs are different every time.
* API Documenation using docstrings
  * Metropolis algorithm methods
  * Data methods and containers
  * Distribution classes and functions
  * Statistical methods
  * Sampling methods
  * Ising model functions
  * Simulation run methods
* Miscellaneous
  * Ignore most VSCode settings and workspace files

Below are additional details on how the simulation was implemented.

Sampling method for Ising model
* Method to sweep grid, where a sweep means for N sites in lattice, pick and try
  to flip a spin N times.
* Methods to sample a new state by picking a site and test flipping it
* Method to update site if flip is accepted
* Logic to update trace and estimators based on chosen sampling interval

Ising model implementation
* Container for simulation parameters
* Method to compute total energy and energy change in a spin flip
* Method to compute total magnetization and magnetization change in a spin flip.
* Method to test a spin flip in the Ising model
* Methods to compute running averages for the 2nd, 3rd, and 4th moments
  for the energy and magnetization
* Function naming scheme that aims to be general and aligned with naming
  conventions in statistics
* Comprehensive use of type annotations

Pre-simulation, main, and post-simulation methods
* Initialize simulation based on input parameters
* Pre-simulation equilibration runs
* Simulation sweeps
* Post-simulation write output to disk as CSV file, and print energy and
  magnetization estimators
